lwtime/nattime.go

```diff
@@ -0,0 +1,196 @@
+	Previous  NatTimeAdjective = "previous"
+	Last      NatTimeAdjective = "last"
+)
+
+func (nta NatTimeAdjective) IsValid() bool {
+	switch NatTimeAdjective(strings.ToLower(string(nta))) { // inline lowercase conversion
+	case Today, Yesterday, This, Current, Previous, Last:
+		return true
+	}
+	return false
+}
+
+type NatTime struct {
+	Adjective NatTimeAdjective
+	Num       string
+	INum      int
+	Unit      RelTimeUnit
+}
+
+func (nt *NatTime) LoadRelTimeUnit(u string) bool {
+	switch strings.ToLower(u) {
+	case "year", "years":
+		nt.Unit = Year
+	case "month", "months":
+		nt.Unit = Month
+	case "week", "weeks":
+		nt.Unit = Week
+	case "day", "days":
+		nt.Unit = Day
+	case "hour", "hours":
+		nt.Unit = Hour
+	case "minute", "minutes":
+		nt.Unit = Minute
+	case "second", "seconds":
+		nt.Unit = Second
+	default:
+		return false
+	}
+	return true
+}
+
+func (nt *NatTime) Parse(s string) error {
+	s = strings.ToLower(s)
+	// Today
+	if NatTimeAdjective(s) == Today {
+		nt.Adjective = This
+		nt.Num = "1"
+		nt.INum = 1
+		nt.Unit = RelTimeUnit("d")
+		return nil
+	}
+	// Yesterday
+	if NatTimeAdjective(s) == Yesterday {
+		nt.Adjective = Previous
+		nt.Num = "1"
+		nt.INum = 1
+		nt.Unit = RelTimeUnit("d")
+		return nil
+	}
+	// Singular
+	var nt_parts []string
+	singularRE := regexp.MustCompile(natTimeSingularRE)
+	if nt_parts = singularRE.FindStringSubmatch(s); s == "" || nt_parts == nil {
+		// Plural
+		pluralRE := regexp.MustCompile(natTimePluralRE)
+		if nt_parts = pluralRE.FindStringSubmatch(s); s == "" || nt_parts == nil {
+			return errors.New(fmt.Sprintf("natural time (%s) is invalid", s))
+		}
+	}
+	// Adjective
+	nt.Adjective = NatTimeAdjective(nt_parts[1])
+	if !nt.Adjective.IsValid() {
+		// this would indicate a code mismatch between enumerated Adjectives and Regex
+		return errors.New(fmt.Sprintf("invalid adjective for natural time (%s)", s))
+	}
+	// Num
+	nt.Num = nt_parts[2]
+	var err error
+	nt.INum, err = strconv.Atoi(nt.Num)
+	if err != nil {
+		nt.Num = "1"
+		nt.INum = 1
+	}
+	// Unit
+	if ok := nt.LoadRelTimeUnit(nt_parts[3]); !ok {
+		return errors.New(fmt.Sprintf("invalid unit for natural time (%s)", s))
+	}
+	return nil
+}
+
+type NatTimeRange struct {
+	Start string
+	End   string
+}
+
+func (ntr *NatTimeRange) Parse(s string) error {
+	nt := NatTime{}
+	if err := nt.Parse(s); err != nil {
+		return err
+	}
+
+	switch nt.Adjective {
+	case This, Current:
+		ntr.Start = fmt.Sprintf("@%s", nt.Unit)
+		ntr.End = "now"
+	case Previous:
+		ntr.Start = fmt.Sprintf("-1%s@%s", nt.Unit, nt.Unit)
+		ntr.End = fmt.Sprintf("@%s", nt.Unit)
+	case Last:
+		ntr.Start = fmt.Sprintf("-%s%s", nt.Num, nt.Unit)
+		ntr.End = "now"
+	default:
+		// This would indicate a code mismatch between NatTime.Parse() and NatTimeRange.Parse()
+		return errors.New(
+			fmt.Sprintf("invalid adjective for natural time (%s)", s))
+	}
+	return nil
+}
+
+func (ntr NatTimeRange) Range() (start time.Time, end time.Time, err error) {
```

Comment:
What is this function for? Why is not being tested?

Reply:
Testing is being done extensively/exclusively on `Range_()`.  This func has no logic other than to call `Range_()` with the current time.  I can add a simple test for this...